src/ducks/account/services.js

```diff
@@ -0,0 +1,91 @@
+import logger from 'cozy-logger'
+import { updateSettings, fetchSettings } from 'ducks/settings/helpers'
+import { ACCOUNT_DOCTYPE, CONTACT_DOCTYPE } from 'doctypes'
+import get from 'lodash/get'
+import set from 'lodash/set'
+import mergeSets from 'utils/mergeSets'
+
+const log = logger.namespace('link-myself-to-accounts')
+
+const fetchMyself = async client => {
+  const response = await client.query(
+    client.find(CONTACT_DOCTYPE, { me: true }).limitBy(1)
+  )
+
+  const [myself] = response.data
+
+  return myself
+}
+
+export const linkMyselfToAccounts = async ({ client }) => {
+  const settings = await fetchSettings(client)
+  const accounts = await client.queryAll(client.all(ACCOUNT_DOCTYPE))
+
+  const alreadyProcessed = new Set(
+    settings.linkMyselfToAccounts.processedAccounts
+  )
+
+  const accountsToProcess = accounts.filter(
+    account => !alreadyProcessed.has(account._id)
+  )
+
+  if (accountsToProcess.length === 0) {
+    log('info', 'No accounts to process, bailing out.')
+    return
+  }
+
+  const myself = await fetchMyself(client)
+
+  if (!myself) {
+    log('info', 'No myself contact found, bailing out.')
+    return
+  }
+
+  for (const account of accountsToProcess) {
+    const currentOwners = get(account, 'relationships.owners.data', [])
+    const newOwners = [
+      ...currentOwners,
+      { _id: myself._id, _type: CONTACT_DOCTYPE }
+    ]
```

Comment:
If settings.linkMyselfToAccounts.processedAccounts is malformed (not correctly saved, or if settings is reset), we risk having the owner twice here. I think we should be a little more defensive here and make sure there are only unique _id here.

```
    const newOwners = uniqueBy([
      ...currentOwners,
      { _id: myself._id, _type: CONTACT_DOCTYPE }
    ], x => x._id)
```

BTW, this logic could be extracted into a `addOwnerToAccount` helper.

Reply:
Thanks, done :+1: